### PR TITLE
fix: walSaveMeta ahead of deleting WAL files as gc in walEndSnapshot

### PR DESCRIPTION
**Pull request by bgzhao66**

### CODE REVIEW

File: source/libs/wal/src/walMgmt.c [hunk 1]
```diff
@@ -121,7 +121,16 @@ SWal *walOpen(const char *path, SWalCfg *pCfg) {
...
+           pWal->path);
+    goto _err;
+  }
+
+  // init gc
+  pWal->toDeleteFiles = taosArrayInit(8, sizeof(SWalFileInfo));
+  if (pWal->toDeleteFiles == NULL) {
+    wError("vgId:%d, failed to init taosArray of toDeleteFiles due to %s. path:%s", pWal->cfg.vgId, strerror(errno),
+           pWal->path);
     goto _err;
```
Comment (cosmic-jc) [inline source/libs/wal/src/walMgmt.c:134]: totDeleteFiles array is not destroyed

File: source/libs/wal/src/walWrite.c [hunk 2]
```diff
@@ -392,6 +376,26 @@ int32_t walEndSnapshot(SWal *pWal) {
...
+    }
+    walBuildIdxName(pWal, pInfo->firstVer, fnameStr);
+    wDebug("vgId:%d, wal remove file %s", pWal->cfg.vgId, fnameStr);
+    if (taosRemoveFile(fnameStr) < 0 && errno != ENOENT) {
+      ASSERT(0);
+    }
+  }
+  taosArrayClear(pWal->toDeleteFiles);
+
 END:
```
Comment (cosmic-jc) [inline source/libs/wal/src/walWrite.c:394]: will it be better to put array clear after end label?

File: source/libs/wal/src/walWrite.c [hunk 3]
```diff
@@ -350,34 +351,17 @@ int32_t walEndSnapshot(SWal *pWal) {
...
-    }
 
   UPDATE_META:
     // make new array, remove files
-    taosArrayPopFrontBatch(pWal->fileInfoSet, actualDelete);
+    taosArrayPopFrontBatch(pWal->fileInfoSet, deleteCnt);
     if (taosArrayGetSize(pWal->fileInfoSet) == 0) {
       pWal->writeCur = -1;
-      pWal->vers.firstVer = -1;
+      pWal->vers.firstVer = pWal->vers.lastVer + 1;
```
Comment (cosmic-jc) [inline source/libs/wal/src/walWrite.c:359]: if all files are deleted, first version should be set to -1

**Review by cosmic-jc — APPROVED** [review]

File: source/libs/wal/src/walMgmt.c (hunk 1 above)
Comment (bgzhao66) [reply source/libs/wal/src/walMgmt.c:134]: resolved.

File: source/libs/wal/src/walWrite.c (hunk 2 above)
Comment (bgzhao66) [reply source/libs/wal/src/walWrite.c:394]: Next time, the partial cleaned WAL files as gc are to be cleaned in next attempt.

File: source/libs/wal/src/walWrite.c (hunk 3 above)
Comment (bgzhao66) [reply source/libs/wal/src/walWrite.c:359]: the close-open interval [) is for an empty set.

**Review by guanshengliang — APPROVED** [review]

**Pull request merged**